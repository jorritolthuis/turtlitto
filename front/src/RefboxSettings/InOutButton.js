import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

/**
 * Constructs two buttons for in and out
 * Author: G.W. van der Heijden
 * Author: T.T.P. Franken
 *
 * Props:
 * - onClick: a function on what to do when the button is pressed
 */
const InOutButton = props => {
  return (
    <Container>
      <GoButton
        onClick={() => {
          props.onClick("in");
        }}
      >
        {"Go in"}
      </GoButton>
      <GoButton
        onClick={() => {
          props.onClick("out");
        }}
      >
        {"Go out"}
      </GoButton>
    </Container>
  );
};

const Container = styled.div`
  width: 50%;
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

const GoButton = styled.button`
  background-color: ${props => props.theme.inOutButton};
  height: 4rem;
  width: 75%;
  margin-left: auto;
  margin-right: auto;
  border: 0.25rem solid;
  font-size: 2rem;
  display: block;
  &:active {
    background-color: ${props => props.theme.inOutButtonActive};
  }
`;

InOutButton.propType = {
  onClick: PropTypes.func.isRequired
};

export default InOutButton;
