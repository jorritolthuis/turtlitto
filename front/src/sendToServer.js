export default (message, destination, session) => {
  const l = window.location;
  const msg = JSON.stringify(message);
  console.log(`send ${msg} to ${l.protocol}//${l.host}/api/v1/${destination}`);
  return fetch(`${l.protocol}//${l.host}/api/v1/${destination}`, {
    method: "POST",
    headers: new Headers({
      Authorization: "Basic " + btoa(`user:${session}`)
    }),
    body: msg
  });
};
