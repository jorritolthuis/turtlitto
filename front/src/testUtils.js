import React from "react";
import { ThemeProvider } from "styled-components";
import theme from "./theme";
import { shallow, mount } from "enzyme";

export const shallowWithTheme = tree => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallow(tree, { context });
};

export const mountWithTheme = tree => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();

  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes
  });
};
