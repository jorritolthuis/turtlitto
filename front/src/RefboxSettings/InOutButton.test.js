import React from "react";
import InOutButton from "./InOutButton";
import { shallow } from "enzyme";
import sendToServer from "../sendToServer";
import { mountWithTheme } from "../testUtils";

describe("InOutButton", () => {
  it("shows two buttons with go in or go out written on it", () => {
    const wrapper = mountWithTheme(<InOutButton onClick={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("When clicked,", () => {
  const realFetch = global.fetch;
  let wrapper = null;
  beforeEach(() => {
    wrapper = shallow(
      <InOutButton
        onClick={prop => {
          sendToServer(prop, "command");
        }}
      />
    );
    const l = window.location;
    global.fetch = jest.fn().mockImplementation((url, params) => {
      expect(url).toBe(`${l.protocol}//${l.host}/api/v1/command`);
      expect(params).toMatchSnapshot();
      return Promise.resolve({ ok: true });
    });
  });

  afterEach(() => {
    global.fetch = realFetch;
  });
  it("the Go In button should pass `go_in` to the 'sendToServer' function", () => {
    wrapper.find("#InOutButton__go-in-button").simulate("click");
  });
  it("the Go Out button should pass `go_out` to the 'sendToServer' function", () => {
    wrapper.find("#InOutButton__go-out-button").simulate("click");
  });
});
