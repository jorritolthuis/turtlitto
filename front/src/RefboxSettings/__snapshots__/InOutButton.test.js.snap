// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InOutButton shows two buttons with go in or go out written on it 1`] = `
.c0 {
  width: 50%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  background-color: #ededed;
  height: 4rem;
  width: 75%;
  margin-left: auto;
  margin-right: auto;
  border: 0.25rem solid;
  font-size: 2rem;
  display: block;
}

.c1:active {
  background-color: #c0c0c0;
}

<InOutButton
  onClick={[Function]}
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.button
        id="InOutButton__go-in-button"
        onClick={[Function]}
      >
        <button
          className="c1"
          id="InOutButton__go-in-button"
          onClick={[Function]}
        >
          Go in
        </button>
      </styled.button>
      <styled.button
        id="InOutButton__go-out-button"
        onClick={[Function]}
      >
        <button
          className="c1"
          id="InOutButton__go-out-button"
          onClick={[Function]}
        >
          Go out
        </button>
      </styled.button>
    </div>
  </styled.div>
</InOutButton>
`;

exports[`When clicked, the Go In button should pass \`go_in\` to the 'sendToServer' function 1`] = `
Object {
  "body": "\\"go_in\\"",
  "headers": Headers {
    "map": Object {
      "authorization": "Basic dXNlcjp1bmRlZmluZWQ=",
    },
  },
  "method": "POST",
}
`;

exports[`When clicked, the Go Out button should pass \`go_out\` to the 'sendToServer' function 1`] = `
Object {
  "body": "\\"go_out\\"",
  "headers": Headers {
    "map": Object {
      "authorization": "Basic dXNlcjp1bmRlZmluZWQ=",
    },
  },
  "method": "POST",
}
`;
